cunumeric/_ufunc/ufunc.py

```diff
@@ -262,37 +321,132 @@ def __call__(
+        out=None,
+        where=True,
+        casting="same_kind",
+        order="K",
+        dtype=None,
+        **kwargs,
+    ):
+        x = convert_to_cunumeric_ndarray(x)
+
+        if out is not None:
+            if not isinstance(out, tuple):
+                raise ValueError("'out' must be a tuple")
+            elif len(out) != self.nout:
+                if len(out) != self.nout:
+                    raise ValueError(f"'out' must be a {self.nout}-tuple")
+
+            # Check if the broadcasting is possible
+            broadcast_shapes(x.shape, *(o.shape for o in out))
+
+        if not isinstance(where, bool) or not where:
+            raise NotImplementedError(
+                "the 'where' keyword is not yet supported"
+            )
+
+        # If no dtype is given to prescribe the accuracy, we use the dtype
+        # of the input
+        precision_fixed = False
+        if dtype is not None:
+            # If a dtype is given, that determines the precision
+            # of the computation.
+            precision_fixed = True
+            x = self._maybe_cast_input(x, dtype, casting)
+
+        # Resolve the dtype to use for the computation and cast the input
+        # if necessary. If the dtype is already fixed by the caller,
+        # the dtype must be one of the dtypes supported by this operation.
+        x, res_dtypes = self._resolve_dtype(x, casting, precision_fixed)
+
+        if out is None:
+            results = tuple(
+                ndarray(shape=x.shape, dtype=res_dtype, inputs=(x, where))
+                for res_dtype in res_dtypes
+            )
+            out = results
+        else:
+
+            def maybe_create_output(out, res_dtype):
+                if out.dtype != res_dtype:
+                    if not np.can_cast(res_dtype, out.dtype, casting=casting):
+                        raise TypeError(
+                            f"Cannot cast ufunc '{self._name}' output from "
+                            f"{res_dtype} to {out.dtype} with casting rule "
+                            f"'{casting}'"
+                        )
+                    return ndarray(
+                        shape=out.shape, dtype=res_dtype, inputs=(x, where)
+                    )
+                else:
+                    return out
+
+            results = tuple(
+                maybe_create_output(o, res_dtype)
+                for o, res_dtype in zip(out, res_dtypes)
+            )
+
+        result_thunks = tuple(result._thunk for result in results)
+        result_thunks[0].unary_op(
+            self._op_code, x._thunk, where, (), multiout=result_thunks[1:]
+        )
+
+        out = tuple(
+            out
+            if out is result
+            else out._thunk.convert(result._thunk, warn=False)
+            for out, result in zip(out, results)
+        )
+
+        return out
+
+
+class binary_ufunc(ufunc):
+    def __init__(
+        self, name, doc, op_code, types, red_code=None, use_common_type=True
```

Comment:
It looks like `use_common_type` is currently `True` for all binary ufuncs, but I think it's supposed to be `False` for `ldexp`.

Reply:
should be fixed now.